hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/impl/listener/ClientNonSmartListenerService.java

```diff
@@ -23,100 +23,134 @@
+    public boolean deregisterListener(String userRegistrationId) {
+        synchronized (listenerRegLock) {
+            ClientRegistrationKey key = new ClientRegistrationKey(userRegistrationId);
+            ClientEventRegistration registration = registrations.get(key);
+
+            if (registration == null) {
+                return false;
+            }
+
+            BaseClientRemoveListenerRequest removeRequest = registration.getRemoveRequest();
+            removeRequest.setRegistrationId(registration.getServerRegistrationId());
+            try {
+                Future future = new ClientInvocation(client, removeRequest).invoke();
+                future.get();
+                removeEventHandler(registration.getCallId());
+                registrations.remove(key);
+            } catch (Exception e) {
+                throw new HazelcastException("Listener with id " + userRegistrationId + " could not be removed", e);
+            }
+            return true;
         }
-        return true;
     }
 
     @Override
-    public void memberAdded(MembershipEvent membershipEvent) {
-        Address ownerConnectionAddress = client.getClientClusterService().getOwnerConnectionAddress();
-        if (membershipEvent.getMember().getAddress().equals(ownerConnectionAddress)) {
-            executionService.executeInternal(new Runnable() {
-                @Override
-                public void run() {
+    public void connectionAdded(Connection connection) {
+        executionService.executeInternal(new Runnable() {
+            @Override
+            public void run() {
+                synchronized (listenerRegLock) {
                     for (ClientRegistrationKey registrationKey : registrations.keySet()) {
-                        invoke(registrationKey);
+                        try {
+                            ClientEventRegistration registration = invoke(registrationKey);
+                            registrations.put(registrationKey, registration);
+                        } catch (Exception e) {
+                            logger.warning("Listener " + registrationKey + " could not be added ");
+                        }
                     }
                 }
-            });
-        }
+            }
+        });
     }
 
     @Override
-    public void memberRemoved(MembershipEvent membershipEvent) {
-        Address ownerConnectionAddress = client.getClientClusterService().getOwnerConnectionAddress();
-        if (membershipEvent.getMember().getAddress().equals(ownerConnectionAddress)) {
-            for (ClientEventRegistration registration : registrations.values()) {
-                removeEventHandler(registration.getCallId());
+    public void connectionRemoved(Connection connection) {
+        synchronized (listenerRegLock) {
+            for (Map.Entry<ClientRegistrationKey, ClientEventRegistration> entry : registrations.entrySet()) {
+                removeEventHandler(entry.getValue().getCallId());
```

Comment:
The lock holding time can also be decreased here by separating the invocations from map removals.